### PR TITLE
Initial flexible folder tag

### PR DESCRIPTION
**Pull request by eaculb**
Added function get_filename_prefix, calls get_filename_prefix in render_and_clean

(Resolved Issue #173 )

### CODE REVIEW

File: xpdan/formatters.py [hunk 1]
```diff
@@ -3,10 +3,25 @@
 from collections import defaultdict
 from pathlib import Path
 
+base_template = (''
```
Comment (CJ-Wright) [inline xpdan/formatters.py]: Why did this get copied over?

**Comment by CJ-Wright** [comment]
Tests are failing due to a Streamz API break which broke SHED. Once the SHED PR goes in the tests should start working again.

**Comment by CJ-Wright** [comment]
Also this PR needs some tests to make certain that everything works as advertised.

File: xpdan/formatters.py [hunk 2]
```diff
@@ -3,10 +3,10 @@
 from collections import defaultdict
 from pathlib import Path
 
-
```
Comment (CJ-Wright) [inline xpdan/formatters.py]: Usually we leave two lines between the function defs and the imports

File: xpdan/formatters.py [hunk 3]
```diff
@@ -65,11 +66,48 @@ def clean_template(template, removals=None, cfmt=cfmt):
...
     f = f.replace('/_', '/')
     return f
 
+# Replaces each parameter name within folder_tag_list with the value of that parameter.
+def get_filename_prefix(folder_tag_list,md):
+    result =''
+    for tag in folder_tag_list:
+        if isinstance(tag,str):
+            try:
+                raw_addition = md[tag]
```
Comment (CJ-Wright) [inline xpdan/formatters.py]: you might want to use the `dict.get(key, default)` syntax so you won't need the try except.
```python
raw_addition = md.get(tag, '')
```

File: xpdan/formatters.py [hunk 4]
```diff
@@ -65,11 +66,48 @@ def clean_template(template, removals=None, cfmt=cfmt):
...
+    e = e.replace('\'','')
     f = Path(e).as_posix()
     f = f.replace('/_', '/')
     return f
 
+# Replaces each parameter name within folder_tag_list with the value of that parameter.
+def get_filename_prefix(folder_tag_list,md):
+    result =''
+    for tag in folder_tag_list:
+        if isinstance(tag,str):
```
Comment (CJ-Wright) [inline xpdan/formatters.py]: space after `,`

File: xpdan/formatters.py [hunk 5]
```diff
@@ -65,11 +66,48 @@ def clean_template(template, removals=None, cfmt=cfmt):
...
+            try:
+                raw_addition = md[tag]
+            except(KeyError):
+                print("Error: " + str(tag) + " not valid folder tag. " + str(tag) + " was skipped in defining the file path.")
+                raw_addition = ''
+        else:
+            sub_md = md.copy()
+            for item in tag:
+                try:
+                    sub_md = sub_md[item]
```
Comment (CJ-Wright) [inline xpdan/formatters.py]: dict get syntax

File: xpdan/tests/test_formatters.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+# See LICENSE.txt for license information.
+#
+##############################################################################
+import os
+
+import pytest
+from xpdan.formatters import (clean_template, get_filename_prefix, render_and_clean)
+import yaml
+
+def get_md_example(filename):
```
Comment (CJ-Wright) [inline xpdan/tests/test_formatters.py]: Please use 4 space intent.

Comment (eaculb) [reply xpdan/tests/test_formatters.py]: Happy to fix this but I'm not seeing where this isn't done -- where do I not 4 space indent? (Still getting used to this IDE so I don't doubt I messed it up in a few places)

**Comment by eaculb** [comment]
@CJ-Wright build failing because of Bluesky API changes

**Comment by CJ-Wright** [comment]
This seems to be fixed in upstream please rebase onto master.
https://git-scm.com/book/en/v2/Git-Branching-Rebasing
https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

File: xpdan/formatters.py [hunk 7]
```diff
@@ -55,22 +57,43 @@ def format_field(self, value, spec):
...
-    z = re.sub(r"__+", "_", d)
-    z = z.replace('_.', '.')
-    e = z.replace('[', '')
-    e = e.replace(']', '')
-    e = e.replace('(', '')
-    e = e.replace(')', '')
-    f = Path(e).as_posix()
+    z = re.sub(r'_+', '_', d)
+    z = z.replace('_.', '.') 
+    z = re.sub(r'[\[\]\(\)\']', '', z)
```
Comment (CJ-Wright) [inline xpdan/formatters.py]: awesome!

**Comment by eaculb** [comment]
@CJ-Wright in the case where a folder list is provided but none of the provided keys exist, I currently have it default back to the sample name as the (only) folder name. Do you think that's ok as a default?

**Pull request merged**